metricflow/dataflow/optimizer/predicate_pushdown_optimizer.py

```diff
@@ -406,31 +422,63 @@ def visit_join_over_time_range_node(self, node: JoinOverTimeRangeNode) -> Optimi
     def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode) -> OptimizeBranchResult:
-        """Updates time range constraint window to account for time spine join intervals, as needed.
+        """Updates pushdown state to account for time spine join types and constraint window requirements.
 
-        For the time being we simply pass through in all cases, because time constraint handling is done between the
-        DataflowPlanBuilder and DataflowToSqlQueryPlanConverter.
+        The JoinToTimeSpineNode is processed as the left node in a join against a single input parent, typically
+        a measure or metric input of some kind, but the join type depends on how the JoinToTimeSpineNode is configured.
+        Since we cannot safely push down filters that might allow for nulls, we must update the pushdown state
+        accordingly.
 
-        TODO: move constraint window management to the optimizer
+        TODO: move constraint window management to the optimizer and enable time-based filter propagation
+        as appropriate
         """
         self._log_visit_node_type(node)
-        return self._default_handler(node)
+        current_pushdown_state = self._predicate_pushdown_tracker.last_pushdown_state
+        if node.join_type is SqlJoinType.LEFT_OUTER or node.join_type is SqlJoinType.FULL_OUTER:
+            updated_pushdown_state = PredicatePushdownState(
+                time_range_constraint=None,
+                where_filter_specs=tuple(),
+                pushdown_enabled_types=current_pushdown_state.pushdown_enabled_types,
+            )
+        else:
+            updated_pushdown_state = PredicatePushdownState.without_time_range_constraint(current_pushdown_state)
 
-    # Other nodes - these simply propagate state, as they do not affect predicate pushdown in our context
+        return self._default_handler(node=node, pushdown_state=updated_pushdown_state)
 
-    def visit_aggregate_measures_node(self, node: AggregateMeasuresNode) -> OptimizeBranchResult:  # noqa: D102
+    # Nodes affecting pushdown state for other reasons
+
+    def visit_compute_metrics_node(self, node: ComputeMetricsNode) -> OptimizeBranchResult:
+        """Disables or refines pushdown state for metrics that require custom handling.
+
+        For example, time filters for derived offset or cumulative metrics might require special state
+        handling for filters on time dimension inputs or time range constraint expressions.
+        """
         self._log_visit_node_type(node)
-        return self._default_handler(node)
+        current_pushdown_state = self._predicate_pushdown_tracker.last_pushdown_state
+        if any(metric_spec.has_time_offset for metric_spec in node.metric_specs):
+            # TODO: Allow non-time filters for offset metrics. This is for parity with the original hook preventing
+            # invalid pushdown for offset metrics
+            updated_pushdown_state = PredicatePushdownState.with_pushdown_disabled()
```

Comment:
Without this we push categorical dimension filters past time spine joins. It's safe in conjunction with the join type handling, but it causes a bunch of weird thrash in snapshots so I'll do it as a follow-up.